contracts/core/MatchExecutor.sol

```diff
@@ -0,0 +1,369 @@
+   */
+  function receiveFlashLoan(
+    IERC20[] calldata tokens,
+    uint256[] calldata amounts,
+    uint256[] calldata feeAmounts,
+    bytes calldata data
+  ) external whenNotPaused {
+    require(msg.sender == address(vault), 'only vault can call');
+    MatchExecutorTypes.Batch[] memory batches = abi.decode(data, (MatchExecutorTypes.Batch[]));
+    /**
+     * execute the matches
+     */
+    _executeMatches(batches);
+
+    /**
+     * payback the loan
+     */
+    for (uint256 i = 0; i < tokens.length; ++i) {
+      IERC20 token = tokens[i];
+      uint256 amount = amounts[i];
+      uint256 feeAmount = feeAmounts[i];
+      token.transfer(address(vault), amount + feeAmount);
+    }
+  }
+  
+  function _executeMatches(MatchExecutorTypes.Batch[] memory batches) internal {
+    uint256 numBatches = batches.length;
+    for (uint256 i; i < numBatches; ) {
+      _broker(batches[i].externalFulfillments);
+      _matchOrders(batches[i].matches);
+
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  function _executeMatchesCalldata(MatchExecutorTypes.Batch[] calldata batches) internal {
+    uint256 numBatches = batches.length;
+    for (uint256 i; i < numBatches; ) {
+      _brokerCalldata(batches[i].externalFulfillments);
+      _matchOrdersCalldata(batches[i].matches);
+
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  /**
+   * @notice Function called to execute a batch of matches by calling the exchange contract
+   * @param matches The batch of matches to execute on the exchange
+   */
+  function _matchOrders(MatchExecutorTypes.MatchOrders[] memory matches) internal {
+    uint256 numMatches = matches.length;
+    if (numMatches > 0) {
+      for (uint256 i; i < numMatches; ) {
+        MatchExecutorTypes.MatchOrdersType matchType = matches[i].matchType;
+        if (matchType == MatchExecutorTypes.MatchOrdersType.OneToOneSpecific) {
+          exchange.matchOneToOneOrders(matches[i].buys, matches[i].sells);
+        } else if (matchType == MatchExecutorTypes.MatchOrdersType.OneToOneUnspecific) {
+          exchange.matchOrders(matches[i].buys, matches[i].sells, matches[i].constructs);
+        } else if (matchType == MatchExecutorTypes.MatchOrdersType.OneToMany) {
+          if (matches[i].buys.length == 1) {
+            exchange.matchOneToManyOrders(matches[i].buys[0], matches[i].sells);
+          } else if (matches[i].sells.length == 1) {
+            exchange.matchOneToManyOrders(matches[i].sells[0], matches[i].buys);
+          } else {
+            revert('invalid one to many order');
+          }
+        } else {
+          revert('invalid match type');
+        }
+        unchecked {
+          ++i;
+        }
+      }
+    }
+  }
+
+  /**
+   * @notice Function called to execute a batch of matches by calling the exchange contract
+   * @param matches The batch of matches to execute on the exchange
+   */
+  function _matchOrdersCalldata(MatchExecutorTypes.MatchOrders[] calldata matches) internal {
```

Comment:
this and the above function look the same

Reply:
The difference between these is that one takes batches via calldata instead of memory. In the case where we don't have to take out a flashloan the matches don't have to be byte encoded/decoded causing this data to be stored in memory. 

Is there a better way to do this?